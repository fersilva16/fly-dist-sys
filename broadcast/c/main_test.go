package main

import (
	"gossip-gloomers/testutils"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	maelstrom "github.com/jepsen-io/maelstrom/demo/go"
	"github.com/stretchr/testify/require"
)

func Test(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)

	go main()

	err := client.InitNode("n0", []string{"n0", "n1", "n2", "n3", "n4"})

	require.NoError(err)

	topologyOutput, err := client.RPC(TopologyRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "topology",
			MsgID: 2,
		},

		Topology: map[string][]string{
			"n0": {"n3", "n1"},
			"n1": {"n4", "n2", "n0"},
			"n2": {"n1"},
			"n3": {"n0", "n4"},
			"n4": {"n1", "n3"},
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, topologyOutput)

	broadcastOutput, err := client.RPC(BroadcastRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "broadcast",
			MsgID: 3,
		},

		Message: 1,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, broadcastOutput)

	broadcastOutputN3, err := link.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, broadcastOutputN3)

	err = client.Write(maelstrom.MessageBody{
		Type:      "broadcast_ok",
		InReplyTo: 1,
	})

	require.NoError(err)

	broadcastOutputN1, err := link.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, broadcastOutputN1)

	retryOutputN1, err := link.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, retryOutputN1)

	err = client.Write(maelstrom.MessageBody{
		Type:      "broadcast_ok",
		InReplyTo: 3,
	})

	require.NoError(err)

	readOutput, err := client.RPC(maelstrom.MessageBody{
		Type:  "read",
		MsgID: 4,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, readOutput)
}
