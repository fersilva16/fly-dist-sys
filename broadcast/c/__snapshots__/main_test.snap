
[Test - 1]
{"src":"n0","dest":"c0","body":{"in_reply_to":2,"type":"topology_ok"}}
---

[Test - 2]
{"src":"n0","dest":"c0","body":{"in_reply_to":3,"type":"broadcast_ok"}}
---

[Test - 3]
{"src":"n0","dest":"n1","body":{"message":1,"msg_id":1,"type":"broadcast"}}
---

[Test - 4]
{"src":"n0","dest":"n3","body":{"message":1,"msg_id":2,"type":"broadcast"}}
---

[Test - 5]
{"src":"n0","dest":"c0","body":{"in_reply_to":4,"messages":[1],"type":"read_ok"}}
---
